### PR TITLE
Build search analyzers on shard recovery

### PR DESCRIPTION
**Pull request by mayya-sharipova**
Separate building analyzers into validation and final building stages.

Today we build analyzers on index creation in two places:
1. On the master node in masterService#updateTask thread for verification.
2. On a shard in clusterApplierService#updateTask thread.

Both these threads must avoid long-running or blocking operations. But some analyzer components will involve long-running or blocking operations such as from an index (e.g. synonym filter). To prevent this, this PR does the following:

1. AnalysisRegistry#build(forValidation, IndexSettings) method is introduced that builds analyzers for validation only and avoids any long-running operations. This method is used on index creation in the above mentioned masterService and clusterApplier threads.
2. Analyzer components beginning with TokenFilterFactory are modified to include `forValidation` parameter that tell them to build a component for validation only.
3. The long-running or blocking operations for building analyzers are moved on index shard creation that is run in the general thread pool (calling `IndexAnalyzers::finalizeBuilding` method).

### CODE REVIEW

**Comment by mayya-sharipova** [comment]
@elastic/es-distributed I am not happy how with this PR if there are problems with loading synonyms, shard recovery will fail, which means an index will be created, but shards will be unassigned (with allocation explain message below).

Would appreciate your help for suggestions.


```txt
{
    "note": "No shard was specified in the explain API request, so this response explains a randomly chosen unassigned shard. There may be other unassigned shards in this cluster which cannot be assigned for different reasons. It may not be possible to assign this shard until one of the other shards is assigned correctly. To explain the allocation of other shards (whether assigned or unassigned) you must specify the target shard in the request to this API.",
    "index": "synonyms4",
    "shard": 0,
    "primary": true,
    "current_state": "unassigned",
    "unassigned_info": {
        "reason": "ALLOCATION_FAILED",
        "at": "2023-05-15T19:27:58.190Z",
        "failed_allocation_attempts": 5,
        "details": "failed shard on node [IOMdhIWHSymej_1yMgH7wA]: failed recovery, failure org.elasticsearch.indices.recovery.RecoveryFailedException: [synonyms4][0]: Recovery failed on {runTask-0}{IOMdhIWHSymej_1yMgH7wA}{uM9eeiEmRGaHP3StP6HJ_Q}{runTask-0}{127.0.0.1}{127.0.0.1:9300}{cdfhilmrstw}{8.9.0}{testattr=test, ml.allocated_processors_double=10.0, ml.machine_memory=68719476736, xpack.installed=true, ml.max_jvm_size=536870912, ml.allocated_processors=10}\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.lambda$executeRecovery$23(IndexShard.java:3147)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:59)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:166)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.StoreRecovery.lambda$recoveryListener$8(StoreRecovery.java:401)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:59)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:166)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:59)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:71)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:27)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:92)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.StoreRecovery.lambda$internalRecoverFromStore$10(StoreRecovery.java:488)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingFailureActionListener.onResponse(ActionListenerImplementations.java:151)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:122)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.CompositeIndexEventListener.iterateBeforeIndexShardRecovery(CompositeIndexEventListener.java:280)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:289)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1662)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.StoreRecovery.internalRecoverFromStore(StoreRecovery.java:411)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.StoreRecovery.recoverFromStore(StoreRecovery.java:93)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.recoverFromStore(IndexShard.java:2288)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:72)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)\n\tat java.base/java.lang.Thread.run(Thread.java:1623)\nCaused by: [synonyms4/DdAR7ZJQTBW7_jj9p1oaEQ][[synonyms4][0]] org.elasticsearch.index.shard.IndexShardRecoveryException: failed recovery\n\t... 23 more\nCaused by: java.lang.IllegalArgumentException: IOException while reading synonyms_path_path: /Users/mayya/Elastic/elasticsearch/build/testclusters/runTask-0/config/synonyms2.txt\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.analysis.Analysis.getReaderFromFile(Analysis.java:291)\n\tat org.elasticsearch.analysis.common@8.9.0-SNAPSHOT/org.elasticsearch.analysis.common.SynonymTokenFilterFactory.getRulesFromSettings(SynonymTokenFilterFactory.java:156)\n\tat org.elasticsearch.analysis.common@8.9.0-SNAPSHOT/org.elasticsearch.analysis.common.SynonymTokenFilterFactory.getChainAwareTokenFilterFactory(SynonymTokenFilterFactory.java:84)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.analysis.AnalyzerComponents.createComponents(AnalyzerComponents.java:81)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.analysis.ReloadableCustomAnalyzer.reload(ReloadableCustomAnalyzer.java:116)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.analysis.IndexAnalyzers$1.reload(IndexAnalyzers.java:157)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.mapper.MapperService.reloadSearchAnalyzers(MapperService.java:536)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.newEngineConfig(IndexShard.java:3276)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.innerOpenEngineAndTranslog(IndexShard.java:1957)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.openEngineAndRecoverFromTranslog(IndexShard.java:1936)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.StoreRecovery.lambda$internalRecoverFromStore$9(StoreRecovery.java:479)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$1.doRun(ActionRunnable.java:34)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\t... 15 more\nCaused by: java.nio.file.NoSuchFileException: /Users/mayya/Elastic/elasticsearch/build/testclusters/runTask-0/config/synonyms2.txt\n\tat java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)\n\tat java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)\n\tat java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)\n\tat java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:261)\n\tat java.base/java.nio.file.Files.newByteChannel(Files.java:379)\n\tat java.base/java.nio.file.Files.newByteChannel(Files.java:431)\n\tat java.base/java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:422)\n\tat java.base/java.nio.file.Files.newInputStream(Files.java:159)\n\tat java.base/java.nio.file.Files.newBufferedReader(Files.java:2892)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.analysis.Analysis.getReaderFromFile(Analysis.java:280)\n\t... 27 more\n",
        "last_allocation_status": "no"
    },
    "can_allocate": "no",
    "allocate_explanation": "Elasticsearch isn't allowed to allocate this shard to any of the nodes in the cluster. Choose a node to which you expect this shard to be allocated, find this node in the node-by-node explanation, and address the reasons which prevent Elasticsearch from allocating this shard there.",
    "node_allocation_decisions": [
        {
            "node_id": "IOMdhIWHSymej_1yMgH7wA",
            "node_name": "runTask-0",
            "transport_address": "127.0.0.1:9300",
            "node_attributes": {
                "testattr": "test",
                "ml.allocated_processors_double": "10.0",
                "ml.machine_memory": "68719476736",
                "xpack.installed": "true",
                "ml.max_jvm_size": "536870912",
                "ml.allocated_processors": "10"
            },
            "node_decision": "no",
            "weight_ranking": 1,
            "deciders": [
                {
                    "decider": "max_retry",
                    "decision": "NO",
                    "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [POST /_cluster/reroute?retry_failed&metric=none] to retry, [unassigned_info[[reason=ALLOCATION_FAILED], at[2023-05-15T19:27:58.190Z], failed_attempts[5], failed_nodes[[IOMdhIWHSymej_1yMgH7wA]], delayed=false, last_node[IOMdhIWHSymej_1yMgH7wA], details[failed shard on node [IOMdhIWHSymej_1yMgH7wA]: failed recovery, failure org.elasticsearch.indices.recovery.RecoveryFailedException: [synonyms4][0]: Recovery failed on {runTask-0}{IOMdhIWHSymej_1yMgH7wA}{uM9eeiEmRGaHP3StP6HJ_Q}{runTask-0}{127.0.0.1}{127.0.0.1:9300}{cdfhilmrstw}{8.9.0}{testattr=test, ml.allocated_processors_double=10.0, ml.machine_memory=68719476736, xpack.installed=true, ml.max_jvm_size=536870912, ml.allocated_processors=10}\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.lambda$executeRecovery$23(IndexShard.java:3147)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:59)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:166)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.StoreRecovery.lambda$recoveryListener$8(StoreRecovery.java:401)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:59)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:166)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:59)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:71)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:27)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:92)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.StoreRecovery.lambda$internalRecoverFromStore$10(StoreRecovery.java:488)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingFailureActionListener.onResponse(ActionListenerImplementations.java:151)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:122)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.CompositeIndexEventListener.iterateBeforeIndexShardRecovery(CompositeIndexEventListener.java:280)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.CompositeIndexEventListener.beforeIndexShardRecovery(CompositeIndexEventListener.java:289)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.preRecovery(IndexShard.java:1662)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.StoreRecovery.internalRecoverFromStore(StoreRecovery.java:411)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.StoreRecovery.recoverFromStore(StoreRecovery.java:93)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.recoverFromStore(IndexShard.java:2288)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:72)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)\n\tat java.base/java.lang.Thread.run(Thread.java:1623)\nCaused by: [synonyms4/DdAR7ZJQTBW7_jj9p1oaEQ][[synonyms4][0]] org.elasticsearch.index.shard.IndexShardRecoveryException: failed recovery\n\t... 23 more\nCaused by: java.lang.IllegalArgumentException: IOException while reading synonyms_path_path: /Users/mayya/Elastic/elasticsearch/build/testclusters/runTask-0/config/synonyms2.txt\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.analysis.Analysis.getReaderFromFile(Analysis.java:291)\n\tat org.elasticsearch.analysis.common@8.9.0-SNAPSHOT/org.elasticsearch.analysis.common.SynonymTokenFilterFactory.getRulesFromSettings(SynonymTokenFilterFactory.java:156)\n\tat org.elasticsearch.analysis.common@8.9.0-SNAPSHOT/org.elasticsearch.analysis.common.SynonymTokenFilterFactory.getChainAwareTokenFilterFactory(SynonymTokenFilterFactory.java:84)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.analysis.AnalyzerComponents.createComponents(AnalyzerComponents.java:81)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.analysis.ReloadableCustomAnalyzer.reload(ReloadableCustomAnalyzer.java:116)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.analysis.IndexAnalyzers$1.reload(IndexAnalyzers.java:157)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.mapper.MapperService.reloadSearchAnalyzers(MapperService.java:536)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.newEngineConfig(IndexShard.java:3276)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.innerOpenEngineAndTranslog(IndexShard.java:1957)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.IndexShard.openEngineAndRecoverFromTranslog(IndexShard.java:1936)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.shard.StoreRecovery.lambda$internalRecoverFromStore$9(StoreRecovery.java:479)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$1.doRun(ActionRunnable.java:34)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\t... 15 more\nCaused by: java.nio.file.NoSuchFileException: /Users/mayya/Elastic/elasticsearch/build/testclusters/runTask-0/config/synonyms2.txt\n\tat java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)\n\tat java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)\n\tat java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)\n\tat java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:261)\n\tat java.base/java.nio.file.Files.newByteChannel(Files.java:379)\n\tat java.base/java.nio.file.Files.newByteChannel(Files.java:431)\n\tat java.base/java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:422)\n\tat java.base/java.nio.file.Files.newInputStream(Files.java:159)\n\tat java.base/java.nio.file.Files.newBufferedReader(Files.java:2892)\n\tat org.elasticsearch.server@8.9.0-SNAPSHOT/org.elasticsearch.index.analysis.Analysis.getReaderFromFile(Analysis.java:280)\n\t... 27 more\n], allocation_status[deciders_no]]]"
                }
            ]
        }
    ]
}
```

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -3265,6 +3265,11 @@ private EngineConfig newEngineConfig(LongSupplier globalCheckpointSupplier) {
             }
         };
         final boolean isTimeBasedIndex = mapperService == null ? false : mapperService.mappingLookup().hasTimestampField();
+        if (mapperService != null) {
+            // we build analyzers that require longer running operations and blocking calls
+            // here during shard recovery in the generic thread pool.
+            mapperService.reloadSearchAnalyzers();
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:3271]: We are holding various mutexes when we get here, and we also get here on certain primary failovers, so this feels like the wrong place to do this. But admittedly I don't know what the exact dependencies are. Can we move it earlier (or later) in the recovery process? For instance could we do this in `org.elasticsearch.index.shard.IndexEventListener#beforeIndexShardRecovery` (which is already a naturally async step of the recovery process)?

**Comment by jimczi** [comment]
>  I am not happy how with this PR if there are problems with loading synonyms, shard recovery will fail, which means an index will be created, but shards will be unassigned (with allocation explain message below).

This is already happening today with file based synonyms. We validate with the file that is present on the master node and we have no guarantee that it's the same or even that it's present on every node. With index based synonyms the validation would be more accurate but without any guarantee that the index will be available when the shard starts or that it was updated in between. 

Although I am a bit on the fence to do this change for existing analyzers since it wasn't caught until now and we have file based synonyms for such a long time. 
So my suggestion would be to add this ability to do `forValidation` but use it only for the new index-based synonym analyzer. This approach is supposed to remove the need for file-based synonyms so it would be good to not introduce a somehow breaking change in the method that we replace. We don't want to introduce another tech debt so for the new method we'd do the right thing which is to not validate on the master node and load the data during shard recovery. 
For file-based synonyms we can continue to do the validation and warn users that they should move to the new system.
Loading the analyzer during shard recovery is the right thing to do so that should be a no-brainer on the data node side.

The piece that I am missing is how we'll make the loading of analyzer async during the shard recovery. It's not needed for file-based synonyms but the problem appears if we want to retrieve the synonyms from the system index. My understanding is that it's not ok to add a blocking call during shard recovery (cc @DaveCTurner) so we need to discuss if/how we can refactor the analyzer loading?

**Comment by DaveCTurner** [comment]
> My understanding is that it's not ok to add a blocking call during shard recovery (cc @DaveCTurner) so we need to discuss if/how we can refactor the analyzer loading?

I'd _prefer_ it to be async, but there's only a limited number of recoveries ongoing at once so we'd only be blocking a few generic threads. And we have these blocking calls today on much more important threads so it's already an improvement to move them to generic threads. I think we can worry about making things more async in a follow-up.

(but if we load the analysers in `IndexEventListener#beforeIndexShardRecovery` then we already have the right API to go async in that followup)

**Comment by jimczi** [comment]
> I think we can worry about making things more async in a follow-up.

Wonderful, thanks for explaining and for giving us a path forward!

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: @DaveCTurner  Thanks for the feedback. Addressed in https://github.com/elastic/elasticsearch/pull/96114/commits/bb367f9fbe385db4ca0ae033e6fd35f49c95abfd

**Comment by mayya-sharipova** [comment]
@DaveCTurner @jimczi Thank you for your feedback. I've tried to address it in https://github.com/elastic/elasticsearch/pull/96114/commits/bb367f9fbe385db4ca0ae033e6fd35f49c95abfd.

**Comment by mayya-sharipova** [comment]
Closing this bigger PR for a smaller temporary solution in PR #96674

**Pull request closed without merging**